File: LeetCode/1793. Maximum Score of a Good Subarray/solution.py
class Solution:
    def maximumScore(self, A, k):
        res = mini = A[k]
        i, j, n = k, k, len(A)
        while i > 0 or j < n - 1:
            if (A[i - 1] if i else 0) < (A[j + 1] if j < n - 1 else 0):
                j += 1
            else:
                i -= 1
            mini = min(mini, A[i], A[j])
            res = max(res, mini * (j - i + 1))
        return res        

class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        n = len(nums)
        left, right = [i for i in range(n)], [i for i in range(n)]
        st = []
        for i in range(n):
            while st and nums[st[-1]] >= nums[i]:
                st.pop()
            if not st:
                left[i] = 0
            else:
                left[i] = st[-1] + 1
            st.append(i)
        st = []
        for i in range(n)[::-1]:
            while st and nums[st[-1]] >= nums[i]:
                st.pop()
            if not st:
                right[i] = n-1
            else:
                right[i] = st[-1] - 1
            st.append(i)
        res = 0
        for i in range(n):
            if left[i] <= k <= right[i]:
                res = max(res, nums[i]*(right[i]-left[i]+1))
        return res 

File: LeetCode/1793. Maximum Score of a Good Subarray/solution.cpp
class Solution {
public:
    int maximumScore(vector<int>& nums, int k) {
        int n = nums.size(), ans = 0;
        for (int i = nums[k], l = k, r = k; i >= 1; --i) {
            while (l > 0 && nums[l - 1] >= i) --l;
            while (r + 1 < n && nums[r + 1] >= i) ++r;
            ans = max(ans, (r - l + 1) * i);
        }
        return ans;
    }
};


class Solution {
public:
    int maximumScore(vector<int>& nums, int k) {
        int n = nums.size(), res = 0;
        vector<int> left(n), right(n);
        for (int i = 0; i < n; i++){
            left[i] = i;
            right[i] = i;
        }
        vector<int> st;
        for (int i = 0; i < n; i++){
            while (!st.empty() && nums[st.back()] >= nums[i]){
                st.pop_back();
            }
            if (st.empty())
                left[i] = 0;
            else
                left[i] = st.back()+1;
            // cout << left[i] << endl;
            st.push_back(i);
        }
        st.clear();
        for (int i = n-1; i >= 0; i--){
            while (!st.empty() && nums[st.back()] >= nums[i]){
                st.pop_back();
            }
            if (st.empty())
                right[i] = n-1;
            else
                right[i] = st.back()-1;
            // cout << right[i] << endl;
            st.push_back(i);
        }
        for (int i = 0; i < n; i++)
            if (left[i] <= k && k <= right[i])
                res = max(res, nums[i]*(right[i]-left[i]+1));
        return res;

    }
};